File: Strivers-A2Z-DSA-Sheet/03.Strings/2.Medium/javascript/04.Implement_atoi.js
/*
============================================================
Problem: String to Integer (atoi) (LeetCode 8)
============================================================
Implement myAtoi(string s) which converts a string to a 32-bit signed integer.

Example 1: Input: "42"       Output: 42
Example 2: Input: "   -42"   Output: -42
Example 3: Input: "4193 with words"  Output: 4193

============================================================
APPROACH 1: BRUTE FORCE - Parse Step by Step
============================================================
Dry Run: s = "   -42"
  Skip spaces → i=3
  s[3]='-' → sign=-1, i=4
  s[4]='4' → ans=4, i=5
  s[5]='2' → ans=42, i=6
  ans * sign = -42
  In range → return -42  ✓

Time: O(N) | Space: O(1)
*/

function myAtoi_BruteForce(s) {
    const INT_MAX = 2147483647, INT_MIN = -2147483648;
    let i = 0, n = s.length;

    // Skip whitespace
    while (i < n && s[i] === ' ') i++;
    if (i === n) return 0;

    // Sign
    let sign = 1;
    if (s[i] === '-') { sign = -1; i++; }
    else if (s[i] === '+') { i++; }

    // Read digits
    let ans = 0;
    while (i < n && s[i] >= '0' && s[i] <= '9') {
        ans = ans * 10 + (s.charCodeAt(i) - 48);
        if (sign === 1 && ans >= INT_MAX) return INT_MAX;
        if (sign === -1 && ans * -1 <= INT_MIN) return INT_MIN;
        i++;
    }

    return ans * sign;
}

/*
============================================================
APPROACH 2: OPTIMAL - Same with early overflow detection
============================================================
Dry Run: s = "2147483648"
  Digits: 2,1,4,7,4,8,3,6,4,8
  ans builds: 2→21→214→2147→21474→214748→2147483→21474836→214748364
  Next digit=8 → 214748364 * 10 + 8 = 2147483648 > INT_MAX → return INT_MAX  ✓

Time: O(N) | Space: O(1)
*/

function myAtoi_Optimal(s) {
    const INT_MAX = 2147483647, INT_MIN = -2147483648;
    let i = 0, n = s.length;

    while (i < n && s[i] === ' ') i++;
    if (i === n) return 0;

    let sign = 1;
    if (s[i] === '-') { sign = -1; i++; }
    else if (s[i] === '+') { i++; }

    let ans = 0;
    const limit = Math.floor(INT_MAX / 10);

    while (i < n && s[i] >= '0' && s[i] <= '9') {
        let digit = s.charCodeAt(i) - 48;

        if (ans > limit || (ans === limit && digit > 7)) {
            return sign === 1 ? INT_MAX : INT_MIN;
        }

        ans = ans * 10 + digit;
        i++;
    }

    return ans * sign;
}

File: Strivers-A2Z-DSA-Sheet/03.Strings/2.Medium/python/04.Implement_atoi.py
"""
============================================================
Problem: String to Integer (atoi) (LeetCode 8)
============================================================
Implement myAtoi(string s) which converts a string to a 32-bit signed integer.

Example 1: Input: "42"       Output: 42
Example 2: Input: "   -42"   Output: -42
Example 3: Input: "4193 with words"  Output: 4193

============================================================
APPROACH 1: BRUTE FORCE - Parse Step by Step
============================================================
Approach:
1. Skip leading whitespace.
2. Check for optional '+'/'-' sign.
3. Read consecutive digits, building the number.
4. Apply sign, then clamp to [INT_MIN, INT_MAX] range.
5. Return the result.

Dry Run: s = "   -42"
  Skip spaces → i=3
  s[3]='-' → sign=-1, i=4
  s[4]='4' → ans=4
  s[5]='2' → ans=42
  return -42  ✓

Time: O(N) | Space: O(1)
"""

INT_MAX = 2**31 - 1
INT_MIN = -(2**31)


def myAtoi_BruteForce(s: str) -> int:
    i, n = 0, len(s)

    # Skip whitespace
    while i < n and s[i] == ' ':
        i += 1
    if i == n:
        return 0

    # Sign
    sign = 1
    if s[i] == '-':
        sign = -1
        i += 1
    elif s[i] == '+':
        i += 1

    # Read digits
    ans = 0
    while i < n and s[i].isdigit():
        ans = ans * 10 + int(s[i])
        i += 1

    ans *= sign
    # Clamp
    if ans > INT_MAX:
        return INT_MAX
    if ans < INT_MIN:
        return INT_MIN
    return ans


"""
============================================================
APPROACH 2: OPTIMAL - Early overflow detection (no big int)
============================================================
Approach:
1. Same steps: whitespace → sign → digits.
2. Before multiplying, check if ans > INT_MAX//10 or (ans == INT_MAX//10 and digit > 7).
3. If overflow, return INT_MAX or INT_MIN immediately.
4. Apply sign and return.

Dry Run: s = "2147483648"
  ans builds: 2→21→214→...→214748364
  Next digit=8: 214748364 > 214748364? No. == and 8>7? Yes → overflow
  Return INT_MAX  ✓

Time: O(N) | Space: O(1)
"""


def myAtoi_Optimal(s: str) -> int:
    i, n = 0, len(s)

    while i < n and s[i] == ' ':
        i += 1
    if i == n:
        return 0

    sign = 1
    if s[i] == '-':
        sign = -1
        i += 1
    elif s[i] == '+':
        i += 1

    ans = 0
    limit = INT_MAX // 10

    while i < n and s[i].isdigit():
        digit = int(s[i])

        if ans > limit or (ans == limit and digit > 7):
            return INT_MAX if sign == 1 else INT_MIN

        ans = ans * 10 + digit
        i += 1

    return ans * sign

